### PR TITLE
handle circular-reference errors in attributes instead of breaking

### LINKED ISSUE #13: Self referential parameters cause a crash in the editor
Eg: Setting a circle's X position to be that same circle's X position. This locks up the whole editor forcing me to refresh the page.

This probably either:
- Shouldn't be allowed, or
- Should have recursion limited

### PR DESCRIPTION
**Pull request by joshuahhh**
Intended to fix #13.

Notes:
- I don't really know how the evaluation / rendering cycle works yet. Experimentation shows that whenever attribute expressions change, `Attribute::dependencies` is always called at some point before `Attribute::value`. I rely upon this, by setting the error state in `Attribute::dependencies` and reading it in `Attribute::value`. I feel uncertain about the consequences of this, especially in edge cases like deserialized models. Let me know if there is a better way of handling this.
- Recursive shapes still seem to work. :smile:
- I updated the value display to include error text. This allows the user to see the cause of the error, rather than just "(Error)".

Screenshot:
error-report

### CODE REVIEW

**Comment by electronicwhisper** [comment]
I agree that relying on `Attribute::value` calling `Attribute::dependencies` and throwing an error from `Attribute::dependencies` is dubious. Maybe a cleaner way would be to add a method `Attribute::hasCircularDependencies` and have `Attribute::value` call this outright in the beginning and throw the error from `Attribute::value`.

Yes, good first step to actually display the error message :smile_cat:

**Comment by joshuahhh** [comment]
@electronicwhisper Does this look any better? (I rearranged the flow in Attribute.coffee and added some tests.)

**Comment by electronicwhisper** [comment]
Hmm. I think I don't like setting that extra state variable `@__circularReferenceError`. There's already quite a bit of these confusing state variables in the Attribute.coffee file. Most of it's for cacheing, which I guess is a necessary evil :frowning:. Ideally there wouldn't be any (if you can think of a way to decouple cacheing that would be cool...).

Maybe a way to get rid of `@__circularReferenceError` would be to make a helper function called like `@_analyzeDependencies` which would return both a list of dependencies and a boolean saying whether it was circular. Then both `@dependencies` and `@checkForCircularReferenceError` would call this helper function.

I'm not sure if it matters much performance-wise, but these lines are at the top of @_value intentionally:

```
# Optimization
if @isNumber()
  return parseFloat(@exprString)
```

Basically almost all attributes are literal values so getting those out of the way first was a big performance boost from when I used to rely on `eval` all the time. Like I said, not sure if it makes a big performance difference in this case, since `@_dependencies` will exit pretty quick, but maybe it should be on top (perhaps with a note). Putting your circular check above this line:

```
referenceValues = _.mapObject @references(), (referenceAttribute) ->
```

would make sense because that is where you will run into trouble if you have a circular reference, so it makes sense to group all that logic together.

I think this `Attribute.coffee` file is one of the ugliest in the codebase. So let's talk about it next time we're together, maybe we can come up with something nicer.

**Comment by joshuahhh** [comment]
OK! Your `@_analyzeDependencies` approach is cleaner. Putting the optimization first also makes sense. PTAL.

(I agree that a more coherent approach to caching would be a really nice improvement. Let's work something out.)
